File: flutter/android/src/main/kotlin/com/spruceid/sprucekit_mobile/Oid4vpAdapter.kt
package com.spruceid.sprucekit_mobile

import android.content.Context
import com.spruceid.mobile.sdk.KeyManager
import com.spruceid.mobile.sdk.rs.DidMethod
import com.spruceid.mobile.sdk.rs.DidMethodUtils
import com.spruceid.mobile.sdk.rs.Oid4vpHolder
import com.spruceid.mobile.sdk.rs.Oid4vpPresentableCredential
import com.spruceid.mobile.sdk.rs.Oid4vpPresentationSigner
import com.spruceid.mobile.sdk.rs.Oid4vpSession
import com.spruceid.mobile.sdk.rs.ParsedCredential
import com.spruceid.mobile.sdk.rs.Oid4vpVersion as RsOid4vpVersion
import com.spruceid.mobile.sdk.rs.Oid4vpResponseOptions as RsResponseOptions
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/**
 * Signer implementation for OID4VP presentation
 */
class Oid4vpSigner(private val keyId: String) : Oid4vpPresentationSigner {
    private val keyManager = KeyManager()
    private val didJwk = DidMethodUtils(DidMethod.JWK)

    // Only the fallback/legacy key is created on demand; a per-credential key
    // must already exist from issuance, else it can't match the cnf binding.
    private fun ensureKey(keyId: String): String {
        val mayGenerate = keyId.isEmpty() || keyId == this.keyId
        val resolvedId = if (mayGenerate) this.keyId.ifEmpty { DEFAULT_KEY_ID } else keyId
        if (!keyManager.keyExists(resolvedId)) {
            require(mayGenerate) {
                "No signing key for per-credential kid '$resolvedId'; it must exist from issuance"
            }
            keyManager.generateSigningKey(id = resolvedId)
        }
        return resolvedId
    }

    private fun jwkFor(keyId: String): String =
        keyManager.getJwk(ensureKey(keyId))?.toString()
            ?: throw IllegalArgumentException("Invalid kid: $keyId")

    override suspend fun sign(keyId: String, payload: ByteArray): ByteArray =
        keyManager.signPayload(ensureKey(keyId), payload)
            ?: throw IllegalStateException("Failed to sign payload")

    override fun algorithm(): String = "ES256"

    override suspend fun verificationMethod(keyId: String): String = didJwk.vmFromJwk(jwkFor(keyId))

    override fun did(keyId: String): String = didJwk.didFromJwk(jwkFor(keyId))

    override fun jwk(keyId: String): String = jwkFor(keyId)

    override fun cryptosuite(): String = "ecdsa-rdfc-2019"

    private companion object {
        const val DEFAULT_KEY_ID = "default_signing_key"
    }
}

/**
 * OID4VP Pigeon Adapter for Android
 *
 * Handles OpenID for Verifiable Presentation flow.
 *
 * Backed by the version-agnostic OID4VP facade (`Oid4vpHolder` /
 * `Oid4vpSession`), which negotiates OID4VP 1.0 or Draft 18 per request.
 * The negotiated version is chosen by the `mode` passed to
 * `handleAuthorizationRequest`.
 */
internal class Oid4vpAdapter(
    private val context: Context,
    private val credentialPackAdapter: CredentialPackAdapter
) : Oid4vp {

    // Session state
    private var holder: Oid4vpHolder? = null
    private var session: Oid4vpSession? = null

    /**
     * Resolves a Dart-side `PresentableCredentialKey` back to the live
     * `Oid4vpPresentableCredential` handle. Built from `session.credentials()`
     * in `handleAuthorizationRequest`, grouped by each credential's `matchId`
     * (the DCQL `credential_query_id` for v1, the input-descriptor id for
     * Draft 18). The same underlying credential may appear under multiple keys
     * when it satisfies multiple queries — those are distinct
     * `Oid4vpPresentableCredential` instances on the Rust side.
     */
    private var credentialsByKey: Map<PresentableCredentialKey, Oid4vpPresentableCredential> = emptyMap()

    /** Maps the pigeon-facing supported versions to the Rust facade enum. */
    private fun rustVersions(versions: List<Oid4vpVersion>): List<RsOid4vpVersion> =
        versions.map { version ->
            when (version) {
                Oid4vpVersion.V1 -> RsOid4vpVersion.V1
                Oid4vpVersion.DRAFT18 -> RsOid4vpVersion.DRAFT18
                Oid4vpVersion.DRAFT13 -> RsOid4vpVersion.DRAFT13
            }
        }

    /**
     * Groups the session's presentable credentials by `matchId`, preserving
     * first-appearance order. Single source of truth for the key map, the flat
     * credential list and the grouped-by-query view.
     */
    private fun groupedByQuery(
        session: Oid4vpSession
    ): List<Pair<String, List<Oid4vpPresentableCredential>>> {
        val map = LinkedHashMap<String, MutableList<Oid4vpPresentableCredential>>()
        for (cred in session.credentials()) {
            map.getOrPut(cred.matchId()) { mutableListOf() }.add(cred)
        }
        return map.map { (qid, creds) -> qid to creds }
    }

    override fun createHolder(
        credentialPackIds: List<String>,
        trustedDids: List<String>,
        keyId: String,
        contextMap: Map<String, String>?,
        callback: (Result<Oid4vpResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                // Get native credentials from packs
                val credentials = mutableListOf<ParsedCredential>()
                for (packId in credentialPackIds) {
                    val packCredentials = credentialPackAdapter.getNativeCredentials(packId)
                    credentials.addAll(packCredentials)
                }

                if (credentials.isEmpty()) {
                    callback(Result.success(Oid4vpError(message = "No credentials found in provided packs")))
                    return@launch
                }

                val signer = Oid4vpSigner(keyId)

                // Create holder (version-agnostic facade)
                val newHolder = Oid4vpHolder.newWithCredentials(
                    credentials,
                    trustedDids,
                    signer,
                    keyId,
                    contextMap,
                    KeyManager()
                )

                synchronized(this@Oid4vpAdapter) {
                    this@Oid4vpAdapter.holder = newHolder
                }

                callback(Result.success(Oid4vpSuccess(message = "Holder created successfully")))
            } catch (e: Exception) {
                callback(Result.success(Oid4vpError(message = e.localizedMessage ?: "Failed to create holder")))
            }
        }
    }

    override fun handleAuthorizationRequest(
        url: String,
        supportedVersions: List<Oid4vpVersion>,
        callback: (Result<HandleAuthRequestResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val currentHolder = synchronized(this@Oid4vpAdapter) { holder }
                if (currentHolder == null) {
                    callback(Result.success(HandleAuthRequestError(
                        message = "Holder not initialized. Call createHolder first."
                    )))
                    return@launch
                }

                // Handle URL format (remove "authorize" if present, similar to Showcase)
                val processedUrl = url.replace("authorize", "")

                // Start a session, restricting negotiation to `supportedVersions`.
                val session = currentHolder.startWithSupportedVersions(processedUrl, rustVersions(supportedVersions))

                // Build (credentialId, matchId) -> credential map and the flat
                // credential list for Dart from a single source: credentials
                // grouped by their `matchId`.
                val groups = groupedByQuery(session)
                val keyMap = mutableMapOf<PresentableCredentialKey, Oid4vpPresentableCredential>()
                val credentialData = mutableListOf<PresentableCredentialData>()
                for ((qid, creds) in groups) {
                    for (cred in creds) {
                        val cid = cred.asParsedCredential().id()
                        val key = PresentableCredentialKey(
                            credentialId = cid,
                            credentialQueryId = qid
                        )
                        keyMap[key] = cred
                        credentialData.add(
                            PresentableCredentialData(
                                credentialId = cid,
                                credentialQueryId = qid,
                                selectiveDisclosable = cred.selectiveDisclosable()
                            )
                        )
                    }
                }

                synchronized(this@Oid4vpAdapter) {
                    this@Oid4vpAdapter.session = session
                    this@Oid4vpAdapter.credentialsByKey = keyMap
                }

                if (credentialData.isEmpty()) {
                    callback(Result.success(HandleAuthRequestError(
                        message = "No matching credentials found for this verification request"
                    )))
                    return@launch
                }

                val info = PermissionRequestInfo(
                    clientId = session.clientId(),
                    domain = session.domain(),
                    purpose = session.purpose(),
                    isMultiCredentialSelection = session.isMultiCredentialSelection(),
                    isMultiCredentialMatching = session.isMultiCredentialMatching()
                )

                callback(Result.success(HandleAuthRequestSuccess(
                    credentials = credentialData,
                    info = info
                )))
            } catch (e: Exception) {
                callback(Result.success(HandleAuthRequestError(
                    message = e.localizedMessage ?: "Failed to handle authorization request"
                )))
            }
        }
    }

    override fun getRequestedFields(key: PresentableCredentialKey): List<RequestedFieldData> {
        val session = synchronized(this) { session } ?: return emptyList()
        val credential = synchronized(this) { credentialsByKey[key] } ?: return emptyList()

        val fields = session.requestedFields(credential)
        return fields.map { field ->
            RequestedFieldData(
                id = field.id,
                name = field.name,
                path = field.path,
                required = field.required,
                retained = field.retained,
                purpose = field.purpose,
                credentialQueryId = field.matchId,
                rawFields = field.rawFields
            )
        }
    }

    override fun submitResponse(
        selectedCredentials: List<PresentableCredentialKey>,
        selectedFieldPaths: List<List<String>>,
        options: ResponseOptions,
        callback: (Result<Oid4vpResult>) -> Unit
    ) {
        CoroutineScope(Dispatchers.IO).launch {
            try {
                val currentSession = synchronized(this@Oid4vpAdapter) { session }
                val keyMap = synchronized(this@Oid4vpAdapter) { credentialsByKey }

                if (currentSession == null) {
                    callback(Result.success(Oid4vpError(message = "Session not initialized")))
                    return@launch
                }

                // Resolve keys to live credential handles
                val resolvedCredentials = selectedCredentials.mapNotNull { keyMap[it] }

                if (resolvedCredentials.isEmpty()) {
                    callback(Result.success(Oid4vpError(message = "No valid credentials selected")))
                    return@launch
                }

                // Create response options. `shouldStripQuotes` and
                // `removeVpPathPrefix` are Draft 18-only knobs not surfaced by
                // the pigeon API; default them off.
                val responseOptions = RsResponseOptions(
                    forceArraySerialization = options.forceArraySerialization,
                    shouldStripQuotes = false,
                    removeVpPathPrefix = false
                )

                // Create and submit the permission response on the session.
                val permissionResponse = currentSession.createPermissionResponse(
                    resolvedCredentials,
                    selectedFieldPaths,
                    responseOptions
                )

                currentSession.submitPermissionResponse(permissionResponse)

                callback(Result.success(Oid4vpSuccess(message = "Presentation submitted successfully")))
            } catch (e: Exception) {
                callback(Result.success(Oid4vpError(
                    message = e.localizedMessage ?: "Failed to submit response"
                )))
            }
        }
    }

    override fun getCredentialRequirements(): List<CredentialRequirementData> {
        val session = synchronized(this) { session } ?: return emptyList()

        val requirements = session.requirements()
        return requirements.map { req ->
            // The facade encodes the requirement's credential query ids as a
            // "|"-joined string in `id`; split it back into the list. Each
            // credential carries its own `matchId`, consistent with the key map
            // built in `handleAuthorizationRequest`.
            val queryIds = req.id.split("|")
            val creds = req.credentials.map { cred ->
                PresentableCredentialData(
                    credentialId = cred.asParsedCredential().id(),
                    credentialQueryId = cred.matchId(),
                    selectiveDisclosable = cred.selectiveDisclosable()
                )
            }
            CredentialRequirementData(
                displayName = req.displayName,
                required = req.required,
                credentialQueryIds = queryIds,
                credentials = creds
            )
        }
    }

    override fun getCredentialsGroupedByQuery(): List<CredentialQueryGroupData> {
        val session = synchronized(this) { session } ?: return emptyList()

        return groupedByQuery(session).map { (qid, creds) ->
            val credentialData = creds.map { cred ->
                PresentableCredentialData(
                    credentialId = cred.asParsedCredential().id(),
                    credentialQueryId = qid,
                    selectiveDisclosable = cred.selectiveDisclosable()
                )
            }
            CredentialQueryGroupData(
                credentialQueryId = qid,
                credentials = credentialData
            )
        }
    }

    override fun getCredentialQueryIds(): List<String> {
        val session = synchronized(this) { session } ?: return emptyList()
        return groupedByQuery(session).map { it.first }
    }

    override fun cancel() {
        synchronized(this) {
            holder = null
            session = null
            credentialsByKey = emptyMap()
        }
    }
}

File: flutter/ios/Classes/Oid4vpAdapter.swift
import Foundation
import SpruceIDMobileSdk
import SpruceIDMobileSdkRs

/// Error types for OID4VP Signer
enum Oid4vpSignerError: Error {
    case illegalArgumentException(reason: String)
}

/// Signer implementation for OID4VP presentation
class Oid4vpSigner: Oid4vpPresentationSigner {
    private let keyId: String
    private let _jwk: String
    private let didJwk = DidMethodUtils(method: SpruceIDMobileSdkRs.DidMethod.jwk)

    init(keyId: String) throws {
        self.keyId = keyId
        if !KeyManager.keyExists(id: keyId) {
            _ = KeyManager.generateSigningKey(id: keyId)
        }
        guard let jwk = KeyManager.getJwk(id: keyId) else {
            throw Oid4vpSignerError.illegalArgumentException(reason: "Invalid kid")
        }
        self._jwk = jwk.description
    }

    func sign(payload: Data) async throws -> Data {
        guard let signature = KeyManager.signPayload(id: keyId, payload: [UInt8](payload)) else {
            throw Oid4vpSignerError.illegalArgumentException(reason: "Failed to sign payload")
        }
        return Data(signature)
    }

    func algorithm() -> String {
        // Parse the jwk as a JSON object and return the "alg" field
        if let data = _jwk.data(using: .utf8),
           let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
           let alg = json["alg"] as? String {
            return alg
        }
        return "ES256"
    }

    func verificationMethod() async -> String {
        do {
            return try await didJwk.vmFromJwk(jwk: _jwk)
        } catch {
            fatalError("Oid4vpSigner: failed to derive verification method from JWK: \(error)")
        }
    }

    func did() -> String {
        do {
            return try didJwk.didFromJwk(jwk: _jwk)
        } catch {
            fatalError("Oid4vpSigner: failed to derive DID from JWK: \(error)")
        }
    }

    func jwk() -> String {
        return _jwk
    }

    func cryptosuite() -> String {
        return "ecdsa-rdfc-2019"
    }
}

/// OID4VP Pigeon Adapter for iOS
///
/// Handles OpenID for Verifiable Presentation flow.
///
/// Backed by the version-agnostic OID4VP facade (`Oid4vpHolder` /
/// `Oid4vpSession`), which negotiates OID4VP 1.0 or Draft 18 per request.
/// The negotiated version is chosen by the `mode` passed to
/// `handleAuthorizationRequest`.
class Oid4vpAdapter: Oid4vp {

    private let credentialPackAdapter: CredentialPackAdapter
    private let lock = NSLock()

    // Session state
    private var holder: Oid4vpHolder?
    private var session: Oid4vpSession?
    /// Resolves a Dart-side `PresentableCredentialKey` back to the live
    /// `Oid4vpPresentableCredential` handle. Built from `session.credentials()`
    /// in `handleAuthorizationRequest`, grouped by each credential's `matchId`
    /// (the DCQL `credential_query_id` for v1, the input-descriptor id for
    /// Draft 18). The same underlying credential may appear under multiple
    /// keys when it satisfies multiple queries — those are distinct
    /// `Oid4vpPresentableCredential` instances on the Rust side.
    private var credentialsByKey: [PresentableCredentialKey: Oid4vpPresentableCredential] = [:]
    /// Dynamic credential offers for the current session, keyed by `offerId`
    /// (surfacing order kept in `dynamicOfferIds`). Dart echoes ids back;
    /// these records are the authoritative ones passed to the Rust session.
    private var dynamicOfferIds: [String] = []
    private var dynamicOffersById: [String: SpruceIDMobileSdkRs.DynamicCredentialOffer] = [:]

    /// Atomically removes and returns the active session while clearing its cached state.
    private func clearSession() -> Oid4vpSession? {
        lock.lock()
        defer { lock.unlock() }
        let activeSession = session
        holder = nil
        session = nil
        credentialsByKey = [:]
        dynamicOfferIds = []
        dynamicOffersById = [:]
        return activeSession
    }

    init(credentialPackAdapter: CredentialPackAdapter) {
        self.credentialPackAdapter = credentialPackAdapter
    }

    /// Maps the pigeon-facing supported versions to the Rust facade enum.
    private func rustVersions(
        _ versions: [Oid4vpVersion]
    ) -> [SpruceIDMobileSdkRs.Oid4vpVersion] {
        versions.map { version in
            switch version {
            case .v1: return .v1
            case .draft18: return .draft18
            case .draft13: return .draft13
            }
        }
    }

    /// Groups the session's presentable credentials by `matchId`, preserving
    /// first-appearance order. Single source of truth for the key map, the
    /// flat credential list and the grouped-by-query view.
    private func groupedByQuery(
        _ session: Oid4vpSession
    ) -> [(qid: String, creds: [Oid4vpPresentableCredential])] {
        var order: [String] = []
        var map: [String: [Oid4vpPresentableCredential]] = [:]
        for cred in session.credentials() {
            let qid = cred.matchId()
            if map[qid] == nil {
                order.append(qid)
                map[qid] = []
            }
            map[qid]?.append(cred)
        }
        return order.map { (qid: $0, creds: map[$0] ?? []) }
    }

    func createHolder(
        credentialPackIds: [String],
        trustedDids: [String],
        keyId: String,
        contextMap: [String: String]?,
        completion: @escaping (Result<Oid4vpResult, any Error>) -> Void
    ) {
        Task {
            do {
                // Get native credentials from packs
                var credentials: [SpruceIDMobileSdkRs.ParsedCredential] = []
                for packId in credentialPackIds {
                    let packCredentials = credentialPackAdapter.getNativeCredentials(packId: packId)
                    credentials.append(contentsOf: packCredentials)
                }

                // Snapshot registered providers for the lifetime of this
                // holder. With providers, an empty credential pack is valid:
                // the whole response may be issued on the fly.
                let providers = SprucekitMobilePlugin.dynamicCredentialProviders

                if credentials.isEmpty && providers.isEmpty {
                    completion(.success(Oid4vpError(message: "No credentials found in provided packs")))
                    return
                }

                // Create signer
                let signer = try Oid4vpSigner(keyId: keyId)

                // Create holder (version-agnostic facade)
                let newHolder = try await Oid4vpHolder.newWithCredentialsAndProviders(
                    providedCredentials: credentials,
                    trustedDids: trustedDids,
                    signer: signer,
                    contextMap: contextMap,
                    keystore: KeyManager(),
                    providers: providers
                )

                lock.lock()
                self.holder = newHolder
                lock.unlock()

                completion(.success(Oid4vpSuccess(message: "Holder created successfully")))
            } catch {
                completion(.success(Oid4vpError(message: error.localizedDescription)))
            }
        }
    }

    func handleAuthorizationRequest(
        url: String,
        supportedVersions: [Oid4vpVersion],
        completion: @escaping (Result<HandleAuthRequestResult, any Error>) -> Void
    ) {
        Task {
            do {
                lock.lock()
                guard let holder = self.holder else {
                    lock.unlock()
                    completion(.success(HandleAuthRequestError(message: "Holder not initialized. Call createHolder first.")))
                    return
                }
                lock.unlock()

                // Handle URL format (remove "authorize" if present, similar to Showcase)
                // A JSON authorization-request body must pass through
                // verbatim, so only munge URL-shaped inputs.
                let trimmed = url.trimmingCharacters(in: .whitespacesAndNewlines)
                let processedUrl = trimmed.hasPrefix("{")
                    ? trimmed
                    : url.replacingOccurrences(of: "authorize", with: "")

                // Start a session, restricting negotiation to `supportedVersions`.
                let session = try await holder.startWithSupportedVersions(
                    request: processedUrl,
                    supportedVersions: rustVersions(supportedVersions)
                )

                // Build (credentialId, matchId) -> credential map and the flat
                // credential list for Dart from a single source: credentials
                // grouped by their `matchId`.
                let groups = groupedByQuery(session)
                var keyMap: [PresentableCredentialKey: Oid4vpPresentableCredential] = [:]
                var credentialData: [PresentableCredentialData] = []
                for group in groups {
                    let qid = group.qid
                    for cred in group.creds {
                        let cid = cred.asParsedCredential().id()
                        let key = PresentableCredentialKey(
                            credentialId: cid,
                            credentialQueryId: qid
                        )
                        keyMap[key] = cred
                        credentialData.append(PresentableCredentialData(
                            credentialId: cid,
                            credentialQueryId: qid,
                            selectiveDisclosable: cred.selectiveDisclosable()
                        ))
                    }
                }

                // Dynamic offers from registered providers; always empty
                // for non-v1 sessions.
                let offers = session.dynamicOffers()

                lock.lock()
                self.session = session
                self.credentialsByKey = keyMap
                self.dynamicOfferIds = offers.map { $0.offerId }
                self.dynamicOffersById = Dictionary(
                    offers.map { ($0.offerId, $0) },
                    uniquingKeysWith: { first, _ in first }
                )
                lock.unlock()

                if credentialData.isEmpty && offers.isEmpty {
                    completion(.success(HandleAuthRequestError(
                        message: "No matching credentials found for this verification request"
                    )))
                    return
                }

                let info = PermissionRequestInfo(
                    clientId: session.clientId(),
                    domain: session.domain(),
                    purpose: session.purpose(),
                    isMultiCredentialSelection: session.isMultiCredentialSelection(),
                    isMultiCredentialMatching: session.isMultiCredentialMatching()
                )

                completion(.success(HandleAuthRequestSuccess(
                    credentials: credentialData,
                    info: info
                )))
            } catch {
                completion(.success(HandleAuthRequestError(message: error.localizedDescription)))
            }
        }
    }

    func getRequestedFields(key: PresentableCredentialKey) throws -> [RequestedFieldData] {
        lock.lock()
        guard let session = self.session,
              let credential = self.credentialsByKey[key] else {
            lock.unlock()
            return []
        }
        lock.unlock()

        let fields = try session.requestedFields(credential: credential)

        return fields.map { field in
            RequestedFieldData(
                id: field.id,
                name: field.name,
                path: field.path,
                required: field.required,
                retained: field.retained,
                purpose: field.purpose,
                credentialQueryId: field.matchId,
                rawFields: field.rawFields
            )
        }
    }

    func submitResponse(
        selectedCredentials: [PresentableCredentialKey],
        selectedFieldPaths: [[String]],
        options: ResponseOptions,
        completion: @escaping (Result<Oid4vpResult, any Error>) -> Void
    ) {
        Task {
            do {
                lock.lock()
                guard let session = self.session else {
                    lock.unlock()
                    completion(.success(Oid4vpError(message: "Session not initialized")))
                    return
                }

                // Resolve keys to live credential handles
                let resolvedCredentials = selectedCredentials.compactMap { self.credentialsByKey[$0] }
                lock.unlock()

                if resolvedCredentials.isEmpty {
                    completion(.success(Oid4vpError(message: "No valid credentials selected")))
                    return
                }

                // Create response options. `shouldStripQuotes` and
                // `removeVpPathPrefix` are Draft 18-only knobs not surfaced by
                // the pigeon API; default them off.
                let responseOptions = SpruceIDMobileSdkRs.Oid4vpResponseOptions(
                    forceArraySerialization: options.forceArraySerialization,
                    shouldStripQuotes: false,
                    removeVpPathPrefix: false
                )

                // Create and submit the permission response on the session.
                let permissionResponse = try await session.createPermissionResponse(
                    selectedCredentials: resolvedCredentials,
                    selectedFields: selectedFieldPaths,
                    responseOptions: responseOptions
                )

                // The verifier's direct_post response may carry a redirect_uri
                // (OID4VP §8.2) sending the user back to the browser.
                let redirectUrl = try await session.submitPermissionResponse(
                    response: permissionResponse)

                completion(
                    .success(
                        Oid4vpSuccess(
                            message: "Presentation submitted successfully",
                            redirectUrl: redirectUrl)))
            } catch {
                completion(.success(Oid4vpError(message: error.localizedDescription)))
            }
        }
    }

    func getCredentialRequirements() throws -> [CredentialRequirementData] {
        lock.lock()
        guard let session = self.session else {
            lock.unlock()
            return []
        }
        lock.unlock()

        let requirements = session.requirements()
        return requirements.map { req in
            // The facade encodes the requirement's credential query ids as a
            // "|"-joined string in `id`; split it back into the list. Each
            // credential carries its own `matchId`, consistent with the key
            // map built in `handleAuthorizationRequest`.
            let queryIds = req.id.split(separator: "|").map(String.init)
            let creds = req.credentials.map { cred -> PresentableCredentialData in
                PresentableCredentialData(
                    credentialId: cred.asParsedCredential().id(),
                    credentialQueryId: cred.matchId(),
                    selectiveDisclosable: cred.selectiveDisclosable()
                )
            }
            return CredentialRequirementData(
                displayName: req.displayName,
                required: req.required,
                credentialQueryIds: queryIds,
                credentials: creds
            )
        }
    }

    func getCredentialsGroupedByQuery() throws -> [CredentialQueryGroupData] {
        lock.lock()
        guard let session = self.session else {
            lock.unlock()
            return []
        }
        lock.unlock()

        return groupedByQuery(session).map { group in
            let creds = group.creds.map { cred in
                PresentableCredentialData(
                    credentialId: cred.asParsedCredential().id(),
                    credentialQueryId: group.qid,
                    selectiveDisclosable: cred.selectiveDisclosable()
                )
            }
            return CredentialQueryGroupData(
                credentialQueryId: group.qid,
                credentials: creds
            )
        }
    }

    func getCredentialQueryIds() throws -> [String] {
        lock.lock()
        guard let session = self.session else {
            lock.unlock()
            return []
        }
        lock.unlock()

        return groupedByQuery(session).map { $0.qid }
    }

    func getDynamicOffers() throws -> [DynamicOfferData] {
        lock.lock()
        let offerIds = self.dynamicOfferIds
        let offersById = self.dynamicOffersById
        lock.unlock()

        return offerIds.compactMap { offersById[$0] }.map { offer in
            DynamicOfferData(
                offerId: offer.offerId,
                credentialQueryId: offer.credentialQueryId,
                title: offer.title
            )
        }
    }

    func submitResponseWithOffers(
        selectedCredentials: [PresentableCredentialKey],
        selectedFieldPaths: [[String]],
        selectedOfferIds: [String],
        options: ResponseOptions,
        completion: @escaping (Result<Oid4vpResult, Error>) -> Void
    ) {
        Task {
            do {
                lock.lock()
                guard let session = self.session else {
                    lock.unlock()
                    completion(.success(Oid4vpError(message: "Session not initialized")))
                    return
                }

                let resolvedCredentials = selectedCredentials.compactMap { self.credentialsByKey[$0] }
                let resolvedOffers = selectedOfferIds.compactMap { self.dynamicOffersById[$0] }
                lock.unlock()

                // Each offer may be selected at most once — a duplicate id
                // would be issued twice and still pass the size check below.
                if Set(selectedOfferIds).count != selectedOfferIds.count {
                    completion(.success(Oid4vpError(message: "Duplicate dynamic offer id selected")))
                    return
                }

                if resolvedOffers.count != selectedOfferIds.count {
                    completion(.success(Oid4vpError(message: "Unknown dynamic offer id selected")))
                    return
                }

                if resolvedCredentials.isEmpty && resolvedOffers.isEmpty {
                    completion(.success(Oid4vpError(message: "No valid credentials or offers selected")))
                    return
                }

                // `shouldStripQuotes` and `removeVpPathPrefix` are Draft
                // 18-only knobs not surfaced by the pigeon API; keep off.
                let responseOptions = SpruceIDMobileSdkRs.Oid4vpResponseOptions(
                    forceArraySerialization: options.forceArraySerialization,
                    shouldStripQuotes: false,
                    removeVpPathPrefix: false
                )

                // Selected offers are issued while creating the response.
                let permissionResponse = try await session.createPermissionResponseWithOffers(
                    selectedCredentials: resolvedCredentials,
                    selectedFields: selectedFieldPaths,
                    selectedOffers: resolvedOffers,
                    responseOptions: responseOptions
                )

                _ = try await session.submitPermissionResponse(response: permissionResponse)

                completion(.success(Oid4vpSuccess(message: "Presentation submitted successfully")))
            } catch {
                completion(.success(Oid4vpError(message: error.localizedDescription)))
            }
        }
    }

    func denyPermission(completion: @escaping (Result<Oid4vpResult, Error>) -> Void) {
        guard let activeSession = clearSession() else {
            completion(
                .success(
                    Oid4vpError(
                        message: "No active OID4VP session"
                    )
                )
            )
            return
        }
        Task {
            do {
                let redirectUrl = try await activeSession.denyPermission()
                completion(
                    .success(
                        Oid4vpSuccess(
                            message: "Permission denied",
                            redirectUrl: redirectUrl
                        )
                    )
                )
            } catch {
                completion(
                    .success(
                        Oid4vpError(
                            message: error.localizedDescription
                        )
                    )
                )
            }
        }
    }

    func cancel() throws {
        clearSession()
    }
}
